enhancements/installer/agent-based-installation-in-hive.md

```diff
@@ -0,0 +1,539 @@
+definition for agent-based installations. Additional CRDs described below will
+be used by other controllers and actors to perform agent-based provisioning.
+
+For the first deliverable, a local deployment of the backing service will
+include new controllers that will operate to achieve the desired state as
+declared in CRDs. Thus the service will expose both REST and Kubernetes-native
+APIs, and both API layers will call the same set of internal methods. The local
+service will include a database; sqlite if possible, or else postgresql.
+
+    --------------------------
+    |  REST API  |  k8s API  |
+    --------------------------
+    | Assisted Service logic |
+    --------------------------
+        |               |
+        v               v
+      SQL DB        File System
+                      storage
+
+For hub clusters that are used for multi-cluster creation and management,
+persistent storage will be a requirement for the database and file system
+storage.
+
+Agents that are running on hardware and waiting for installation-related
+instructions will continue to communicate with the assisted service using the
+existing REST API. In the future that may transition to using a CRD directly.
+In the meantime, status will be propagated to the appropriate k8s resource when
+an agent reports relevant information.
+
+**InstallEnv**
+This new resource, part of the assisted intaller's new operator, represents an
+environment in which a group of hosts share settings related to networking,
+local services, etc. This resource is used to create a discovery image that
+should be used for booting hosts. In the REST API this corresponds to the
+"image" resource that is embedded in the "cluster" resource, but for
+kubernetes-native it is more natural for it to be separate.
+
+The discovery ISO can be downloaded from a URL that is available in the
+resource's Status.
+
+The details of this resource definition were discussed [in a
+pull request](https://github.com/openshift/assisted-service/pull/969).
+
+Summary of fields:
+
+Spec
+* egress proxy settings
+* a list of NTP sources (hostname or IP) to be added to all cluster
+hosts. They are added to any NTP sources that were configured through other
+means.
+* a list of SSH public keys that will be added to all agents for use in debugging.
+* pull secret reference
+* a label selector for Agents. This is how Agent resources are identified as
+belonging to a particular InstallEnv.
+* labels to add to Agent resources upon creation. These labels should match
+the label selector above.
+* ClusterDeployment reference. This is a temporary field that associated an
+InstallEnv with a single cluster definition. This is required because the
+backend assisted service is not capable of producing a discovery ISO unless it
+is scoped to a cluster. That limitation will be removed in future releases.
+
+Status
+* a URL to download the discovery ISO produced for this InstallEnv.
+
+**Agent**
```

Comment:
Similar here, about the only thing I'm still missing in this proposal is what's namespaced and what isn't, and how we maintain multi-tenancy.

Reply:
My recollection was that both Agent and InstallEnv were namespaced, and not tied to specific ClusterDeployments long term. We expect each ClusterDeployment to be in it's own namespace, which means that our label selectors are effectively treated as global. This means that if you grant someone the RBAC to make a ClusterDeployment, they can grab onto any Agents arbitrarily by use of the label selector within their ClusterDeployment, and at present we have no way to control this. At present this would not be a multi-tenant solution.

Probably worth calling out in the sections below as a caveat of some kind.

Reply:
I see. If we have to do label matching across namespaces, that definitely complicates the picture. We'll need to play through some scenarios and maybe adjust this approach to defining relationships.

Reply:
Yeah if InstallEnv is multi-cluster, and Agents are not initially tied to any cluster, and we expect one namespace per ClusterDeployment, the label selector can't be specific to that cluster's namespace. Will get a little tricky, or most likely we just accept that multi-tenancy isn't a thing here.

Reply:
I don't think so, you could still point to any namespace you like on the ClusterDeployment, whether or not you have access. If we move the relationship the something in the InstallEnv ns the same problem exists, you can link to any ClusterDeployment you like. (provided you know or can guess it's name) 

Afaik any namespace boundary crossing will introduce RBAC issues. Not aware of any techniques to work past this but maybe someone has ideas.

Reply:
There is a new section `#### Namespaces and Multi-tenancy` that explains this based on our discussions.